### DISCUSSION TITLE
Design of procedural languages in GOOL

### DISCUSSION THREAD

**Discussion by B-rando1** [JacquesCarette/Drasil#3792] [General]
This discussion started in #3789.  The summary is, I was bringing up an issue I was having with translating GOOL classes into Julia libraries, and @smiths brought up the question of whether this is the right way to go.  The danger of translating OO patterns into a non-OO language is that we might end up with code that is trying to emulate the OO patterns rather than achieving the functionality in a way that is idiomatic in that language.

As I mentioned in the issue, I don't think this is necessarily the case, at the very least not for C code.  The examples I have seen of libraries in procedural languages usually end up looking quite similar to classes anyway, so I don't think this is a bad way to go.

With that said, I might need to do a bit more research on Julia (and potentially Matlab as well) to see if there different patterns that it considers idiomatic, in particular with respect to its more unique features.  The documentation certainly hints at possible patterns, though as of yet I am unsure of whether these patterns are used in production code.  I will do some digging, and I might also write an example 'by hand' if I think that will be helpful.

An example I remembered of a procedural library deviating from the OO way of doing things is GameMaker's particle system.  In GameMaker  to create particles you need a particle system, a particle type, and a particle emitter.  These are separate constructs in GameMaker, so functions often take an instance of each (e.g. [part_emitter_burst](https://manual.gamemaker.io/monthly/en/GameMaker_Language/GML_Reference/Drawing/Particles/Particle_Emitters/part_emitter_burst.htm) needs one of each, as well as the number of particles to generate).  In OO this would likely be done differently, for example with has-a relationships.  Using structs we can still express has-a relationships in procedural languages, but this might serve as evidence that that isn't always the most idiomatic way of doing it.

I guess that my question is, how fragmented do we want to allow GOOL to become?  Currently we have a nice system where we generate one GOOL program, and it cleanly translates to all of the targets (with a few minor exceptions).  Do we really want to push the differences between languages above GOOL and into `drasil-code`?  It seems to me that GOOL would be neither Generic nor OO at that point 😄.  Depending on what I find about Julia and Matlab it might not come to that, but I think it's an important question that we should have an answer ready for.

Some of this is just my thoughts, so what are your thoughts @smiths @JacquesCarette @balacij?  Do you agree that it's still worthwhile to translate GOOL classes into Julia libraries?  Do you think we should investigate generating GOOL code designed specifically for some targets, even at the risk of losing reusability?

**Comment by B-rando1** [comment]
I like this Turtle Graphics example in Julia: https://github.com/JuliaGraphics/Luxor.jl/blob/master/src/Turtle.jl.

Here you see it using multiple constructors.  It doesn't look like they're chaining constructors (they all call the 'true' constructor directly) so this is something that would work in all of our current targets except Python.
```julia
Turtle(x::Float64, y::Float64) = Turtle(x, y, true, 0, (0, 0, 0))
Turtle(pos::Point=O) = Turtle(pos.x, pos.y, true, 0, (0, 0, 0))
Turtle(pos::Point, pendown::Bool) = Turtle(pos.x, pos.y, pendown, 0, (0, 0, 0))
Turtle(pos::Point, pendown::Bool, orientation::Real) = Turtle(pos.x, pos.y, pendown, orientation, (0, 0, 0))
Turtle(pos::Point, pendown::Bool, orientation::Real, col::NTuple{3, Number}) = Turtle(pos.x, pos.y, pendown, orientation, col)
Turtle(pos::Point, pendown::Bool, orientation::Real, r, g, b) = Turtle(pos.x, pos.y, pendown, orientation, (r, g, b))
```

Here you can see the type signature for one of its functions.  All the other functions are very similar as well.
```julia
function Forward(t::Turtle, d=1)
```

I like how straightforward this library is compared to some others.  It's very easy to read, and the API is simple and consistent.  It also demonstrates that a traditional library structure can be used in Julia.

**Reply by JacquesCarette** [reply]
This seems like an interesting form of overloading.

**Reply by smiths** [reply]
I recently read an article about how [Julia solves the Expression problem via multiple dispatch](https://arstechnica.com/science/2020/10/the-unreasonable-effectiveness-of-the-julia-programming-language/)  The multiple dispatch approach isn't really OO or functional.  I need to see some more examples to really get it, but this approach does seem like something different.

From our past conversations, I believe @JacquesCarette is fine if GOOL isn't just for OO.  

I think we should try to follow the approach we've been using of not deciding the "best" option, but instead supporting as many different patterns as we can.  If a certain pattern isn't possible in a given language, then that code cannot be generated.

Since Julia will be able to generate code that is like our OO code, we should generate that code.  In the future, when we understand multiple dispatch better, we could give the option of generating code that follows this paradigm.  The generator in this case wouldn't work for languages that can't be made to support it.

GOOL can grow as new cases come up.  I do think it will be advantageous at some point to find some idiomatic Julia code.  Our current examples are too simple to highlight the differences between languages.  It might help when we get indexing working, since Julia and Matlab are particularly well suited to processing vectors/matrices.

**Reply by B-rando1** [reply]
That was an interesting article @smiths!  This is more of an aside, but it reminded me of Haskell's typeclasses.  I noticed on the Wikipedia [page](https://en.wikipedia.org/wiki/Expression_problem#:~:text=The%20expression%20problem%20is%20a,of%20statically%20typed%20data%20abstractions.) for the Expression problem that the Tagless-final strategy is another solution to the expression problem, so they seem to be similar ideas, at least.

I think you might be right, maybe we should take a step back and focus on what's easiest to translate, and with more time to think about it and look at examples, we can find a better approach to these problems.  It does seem like more of the work will have to be done by `drasil-code` than I initially thought.

**Reply by B-rando1** [reply]
I've been thinking about this some more.  I think that article demonstrates something I hadn't thought of before: it would be beneficial for our code not just to be idiomatic in its own self-contained way, but to allow other programs to interop with it in an idiomatic way.

I don't know if anything has to change to achieve that level of interoperability (it might come 'for free' in Julia), but I think it's a good principle to keep in mind.

**Comment by B-rando1** [comment]
By the way, I started a wiki page about the design decisions related to converting GOOL classes into struct libraries.  [Here](https://github.com/JacquesCarette/Drasil/wiki/Converting-GOOL-Classes-into-Struct-Libraries-for-Procedural-Languages) is the link.  It's currently quite basic and somewhat out-of-date, but I'll do my best to update it.

**Reply by JacquesCarette** [reply]
Thanks for doing this.

**Comment by B-rando1** [comment]
Another place that I should have thought to mention earlier is the Julia standard library!
The [List API](https://docs.julialang.org/en/v1/base/arrays/#Basic-functions) is probably the best place to look.
- Everything under 'Basic functions' follows the 'OO style' of taking an instance of the data structure as its first argument.
- Some of the more complicated functions have a more unique signature, but those are much more specialized functions that don't even have counterparts in most list APIs, so I'm not sure that they're the best examples to go off of.
- The [findfirst](https://docs.julialang.org/en/v1/base/arrays/#Base.findfirst-Tuple{Any}) function is the only real exception I could find.  In one of its implementations, it takes a predicate, then a list, and returns the index of the first element of the list that satisfies the predicate.  I'm not really sure why they chose to put the predicate before the list, it seems like an odd choice to me.
- The API also makes heavy use of multiple dispatch.  This is mainly for user convenience, e.g. the `findfirst` function has an implementation that only takes a list of `Bool`, and returns the index of the first element that is `true`.

**Comment by JacquesCarette** [comment]
First: I don't think we should be trying to 'translate' OO idioms to non-OO languages. It's not that it won't work -- it will work well enough for a while, and then get increasingly hacky, until it feels like it's not working anymore.

What we should be doing is figuring out how to "bundle functionality" together. This belongs in `drasil-code`. There may be a need for a `GenProcLang` (`GPL` is too overloaded, it doesn't bring to mind 'Generic Procedural Language') beside GOOL. They would share a huge amount (i.e. they are both imperative languages at their core).

The question then becomes: where do the OO assumptions come in (in the current code base)? Then we can figure out what to do about it.

**Reply by B-rando1** [reply]
That's a good point about the danger of doing too much in GOOL (as opposed to `drasil-code`), how it will start looking hacky if we take it too far.  It sounds, then, like we want GOOL to translate GOOL code to code with the same meaning (as opposed to the 'equivalent but not the same' meaning I was attempting with translating objects to structs).  Correct me if I misunderstood your comment.

From this, should I stop any work involving objects in procedural languages, and focus solely on translating things that have a 1-to-1 translation?
- If this is the way to go, then I would say we should probably add support for `structs` and/or `hashmaps` to GOOL at some point, as procedural languages would surely benefit from having some way of bundling data together.  This doesn't have to be right away, but we should keep it in mind.

Then from there, we'll need to separate GOOL into the Venn diagram `[OO [OO and Procedural] Procedural]`.  We might not need the Procedural-only part, but definitely will need the OO-only part if we want to benefit from Haskell's type system.  There are a few structures in GOOL that assume a renderer implements all of the OO features, and we'll have to split that up somehow.  I think this is in-line with what you meant when you said we may need a `GenProcLang` beside GOOL.

After that, we'll need to modify `drasil-code` to have more procedural-friendly options.  We could think about having it generate multiple sets of GOOL code for its different targets, so that we don't have the problem we currently have that if your set of `Choices` doesn't work with a target language, you're out of luck.  Alternatively we could have the user hand-pick multiple sets of `Choices` (as in the Projectile example), and generate some languages in one set, and other languages in others.

I think it's probably best (for me at least) to think about this for a few days, so I'll put it on the backburner and we can continue discussing in the meantime.

**Reply by JacquesCarette** [reply]
First paragraph: you understand correctly.

On your next question: I wouldn't quite say 1-to-1 translation. I would say meaning-preserving translation, and definitely only one-way (i.e. from our internal high-level representation to an actual PL).

In theory, `struct`s are degenerate objects, i.e. those with no methods. In practice, it does make sense to have a separate representation for them so that we get this invariant "for free" because it is part of the representation itself. Adding `hashmaps` would indeed be useful.

I agree with the Venn diagram split.

As a fairly large design, it's probably wise to think about it, in the background, over several days.

**Comment by B-rando1** [comment]
I spent some time today tracing out the typeclasses in [ClassInterface.hs](https://github.com/JacquesCarette/Drasil/blob/main/code/drasil-gool/lib/GOOL/Drasil/ClassInterface.hs), trying to get a grasp on how we can start splitting it up into OO-only, OO and Procedural, and Procedural-only typeclasses.  I created a draw.io diagram mapping them out, viewable [here](https://drive.google.com/file/d/19gW8p70-LajaXSgAMJoZtdrW3GHCKhKc/view?usp=drive_link).

I have two things that immediately arise from this:
- First is a question: what is `StatePattern` supposed to do?  From the name it sounds like a design pattern, but it seems to be implemented as a macro for a switch statement.  There's also no examples of it or test cases with it, so I'm not sure what to do with it.
- Second is an observation.  The typeclasses seem to be more tightly coupled to OO than I had hoped.  There are 11 typeclasses on the Procedural side of the diagram that directly mention OO terms, and most of the other typeclasses have a few of them as a dependency.
    - I don't think this is the end of the world, but it'll take some work.  I'm hoping that I can start at the leaves of the dependency tree, splitting typeclasses into their OO and Procedural counterparts.  I think with some luck we'll end up with a new tree where OO stuff only shows up in the OO section.

**Reply by B-rando1** [reply]
I updated the diagram a bit.  The trouble with the diagram is that it needs to summarize 47 typeclasses that are all interconnected, so drawing arrows between them all proved to be infeasible.  Instead I chose to summarize the dependencies of each typeclass by giving its border the colour of its 'most OO' dependency.

The results of this show that every single typeclass in `OOProg` depend on at least one typeclass that is either fully OO or needs to be split up into its fully OO and shared components.

This is not good news of course, but the good part is that `TypeSym` appears to be the culprit behind a good many of these.  Because of this I'd like to discuss it a bit more.

[TypeSym](https://github.com/JacquesCarette/Drasil/blob/d3d569ba519c8fbb6f6b3810faf61a5227e653e3/code/drasil-gool/lib/GOOL/Drasil/ClassInterface.hs#L100-L116) defines the basic types in GOOL.  It is clear by looking at it that all of them except `obj` are shared between OO and procedural.  Because of `obj` I classified it as 'needs to be split up', and since just about everything has a type, this means that all of GOOL relies on something that needs to be split up.

There are two ways we can deal with this:
- Split `TypeSym` up into `OOTypeSym` and `TypeSym`.  I'm not exactly sure how difficult it would be to get this working as it should.  We might be able to follow the example of [VectorType](https://github.com/JacquesCarette/Drasil/blob/d3d569ba519c8fbb6f6b3810faf61a5227e653e3/code/drasil-gool/lib/GOOL/Drasil/ClassInterface.hs#L410-L411) for inspiration, though.  I think splitting `TypeSym` up might have a better end result, as Haskell's type system wouldn't let us create something of type `obj` in a procedural program.
- Keep `TypeSym` as it is, and leave it up to individual renderers to throw errors when a type they don't support is used.  This would be much easier to get going, and would have the benefit of being able to reuse much more existing code and data-structures.  The downside would be that we would have to be a bit more careful when implementing things in procedural languages, as the type system wouldn't stop us from creating something of type `obj`.

I think option 1 is better, so I'll get started on it.  But let me know if you have any additional thoughts, @JacquesCarette.

**Reply by JacquesCarette** [reply]
`StatePattern` is a relic of the dark past when GOOL was created to serve the purposes of SAGA, a story-telling engine generator. That is likely all dead code, and could be deleted.

(I know it's too late but) I agree that option 1 is better.

**Reply by B-rando1** [reply]
Here's an outline of my next steps in this project:
- `FuncAppstatement` and `DeclStatement` should be super easy, as they just need to be separated.
- `StatePattern` can be removed.  I did a quick search and it doesn't look like it's being used anywhere.
- `MethodSym` shouldn't be too bad either.  It needs to be split up, but I think that's all.
- `ModuleSym` is a unique case, as its `buildModule` method is needed by both OO and  Procedural renderers, but it needs a different type signature for each.  The reason for this is that to build OO modules we need to take a list of classes for the module.  I see two options:
    - Do what we've been doing, having the OO typeclass inherit from the procedural one.  In this case there would be the `buildModule` and `buildModuleOO` methods, and OO renderers would need to implement both of them.  This feels a little redundant as we can already generate an OO module with no classes by simply passing an empty list for the classes.  There may be some advantages to having the typeclasses related to each other, though.
    - Have the `ModuleSymOO` and `ModuleSymProc` typeclasses be completely separate.  This has the advantage of removing the redundancy I mentioned above, but since the typeclasses aren't tied to each other we _might_ need to do more work elsewhere to enforce that they are counterparts.
    - I'm not too sure which approach is better.  I'm leaning toward option 2, but they both have pros and cons.

In theory, once the typeclasses I mentioned here are separated, we will have successfully separated GOOL into its OO and shared components.  The next step will be to make `drasil-code` aware of these changes so that it can generate procedural code.  After that we can start adding any Procedural-only features we want.

**Reply by JacquesCarette** [reply]
My feeling is also to go for option 2 - and then see once it works what the consequences are.

**Reply by B-rando1** [reply]
I took a naive crack at separating the 'central' typeclasses in GOOL.  The changes I was making spiraled out of control, and I had to undo the changes.

The issue is that there's a chain of typeclasses that all inherit from `MethodSym` and `ModuleSym`, which both need to be separated.  This means that we end up with ~6 typeclasses that exist 'in parallel' - they do the same things, but the types involves are slightly different.  Then, because `ProgramSym`, the root of this chain, is used in a lot of places, there are many places where decisions have to be made about how to keep everything compatible.

#### Next Steps:
- I think for the most part I just need to take a slower, more structured approach.  This was a bigger problem than I bargained for, and I was just trying to get it done.
- Because these typeclasses exist 'in parallel' rather than one inheriting from the other, I ran into a lot of name conflicts.  If we split the `ClassInterface.hs` file into three files (to match the Venn diagram), we'll be able to give parallel methods and types the same names, rather than giving them ugly prefixes.  This should should help streamline some things.

**Reply by JacquesCarette** [reply]
This "seems simple but spins out of control" is rather common. Glad you spotted it fairly early instead of letting it grow and grow.

Small PRs, some of which might even feel like sideways steps, will get you there.

Splitting things into files seems like a good idea. One can then even use all of them (by importing qualified) if/when needed.

**Reply by B-rando1** [reply]
I had a thought about name conflicts.  Splitting things up into different files will be enough for `drasil-gool`, but then other folders access them through `GOOL.Drasil`, which bundles all of GOOL into one interface.  Because of that, exported names (including the typeclasses we're talking about) still need to be unique, I believe.

Is there a way we could change how we export names that would avoid name conflicts?  There are a good number of typeclasses and functions that will require prefixes in their names otherwise.
- We could split up the file `GOOL.Drasil` into 3 files, each one corresponding to `OO`, `Proc`, and `Common`, respectively.
- We could create two new directories for `OO` and `Proc`.  I think we are planning to do this eventually, just we decided it would be better to get it working before we do too much refactoring.
- There might be other options that I don't know about.

What are your thoughts?

**Reply by JacquesCarette** [reply]
The two new directories seem to be the most future-proof.

**Reply by B-rando1** [reply]
Alright, sounds good.  How much setup is involved to do that?  E.g. editing the `stack.yaml` files, etc.

**Reply by JacquesCarette** [reply]
@balacij can help with the details.

**Comment by B-rando1** [comment]
I was looking at `ScopeSym` (the old one, that has to do with `public` vs `private`).  It's used by `function`, `inOutFunc`, and `docInOutFunc`.  I had assumed that `ScopeSym` should only be used with OO-features, but that doesn't actually seem to be the case.

From experimenting, it looks like the use of `ScopeSym` with functions, similar to with methods, defines whether a function is visible to other modules.
- In Java and C#, functions are implemented as static methods, so this translates in a very obvious way.
- In Swift, functions are declared at the top-level as in Python and C++, but if they are given the `private` keyword, they can only be called from within the module they are in.
- Python and C++ don't do anything with the scope given to functions.

Based on this, I believe my initial assessment was incorrect, and `ScopeSym` should continue to be common between OO and Procedural languages.

**Comment by B-rando1** [comment]
It seems like the main focus(es) of this discussion are finished.  We chose a focus on how to deal with the differences between OO and procedural languages, and were able to make the necessary changes.

There's still a few loose ends, but #3871 and #3904 take care of them for the most part.
